src/MonitorProfiler/MainProfiler/ExceptionTracker.cpp

```diff
@@ -0,0 +1,392 @@
+
+    IfFailRet(GetFullyQualifiedMethodName(moduleId, classId, token, fullMethodName));
+
+    return S_OK;
+}
+
+HRESULT ExceptionTracker::GetFullyQualifiedMethodName(ModuleID moduleId, ClassID classId, mdMethodDef token, tstring& fullMethodName)
+{
+    HRESULT hr = S_OK;
+
+    ComPtr<IMetaDataImport2> metadataImport;
+    IfFailLogRet(_corProfilerInfo->GetModuleMetaData(
+        moduleId,
+        CorOpenFlags::ofRead,
+        IID_IMetaDataImport2,
+        (IUnknown**)&metadataImport
+    ));
+
+    // Get Module Name: typically the full path to the assembly
+    ULONG modulePathCount = 0; // Includes null-terminater
+    IfFailLogRet(_corProfilerInfo->GetModuleInfo(
+        moduleId,
+        nullptr,
+        0,
+        &modulePathCount,
+        nullptr,
+        nullptr
+    ));
+
+    unique_ptr<WCHAR[]> modulePath(new (nothrow) WCHAR[modulePathCount]);
+    IfNullRet(modulePath);
+
+    IfFailLogRet(_corProfilerInfo->GetModuleInfo(
+        moduleId,
+        nullptr,
+        modulePathCount,
+        nullptr,
+        modulePath.get(),
+        nullptr
+    ));
+
+    tstring classNameStr;
+    IfFailRet(GetFullClassName(classId, classNameStr));
+
+    // Get Method Name
+    ULONG methodNameCount = 0; // Includes null-terminater
+    IfFailLogRet(metadataImport->GetMethodProps(
+        token,
+        nullptr,
+        nullptr,
+        0,
+        &methodNameCount,
+        nullptr,
+        nullptr,
+        nullptr,
+        nullptr,
+        nullptr
+    ));
+
+    unique_ptr<WCHAR[]> methodName(new (nothrow) WCHAR[methodNameCount]);
+    IfNullRet(methodName);
+
+    IfFailLogRet(metadataImport->GetMethodProps(
+        token,
+        nullptr,
+        methodName.get(),
+        methodNameCount,
+        nullptr,
+        nullptr,
+        nullptr,
+        nullptr,
+        nullptr,
+        nullptr
+    ));
+
+    // CONSIDER: If generic parameters should be replaced with actual types, do that
+    // as part of calculating the method name. Requires calling GetMethodProps to get
+    // the method signature, parsing the method signature, and resolving type names
+    // from the various type tokens.
+
+    tstring modulePathStr(modulePath.get());
+
+    // The full method name should be in the following format: <ModuleName>!<Namespace>.<TypeName>.<MethodName>
+    // Example: System.Private.CoreLib.dll!System.Collections.Generic.Dictionary`2+Enumerator.MoveNext
+    fullMethodName.clear();
+    fullMethodName.append(modulePathStr.substr(modulePathStr.find_last_of(_T("/\\")) + 1));
+    fullMethodName.append(_T("!"));
+    fullMethodName.append(classNameStr);
+    fullMethodName.append(_T("."));
+    fullMethodName.append(methodName.get());
+
+    return S_OK;
+}
+
+HRESULT ExceptionTracker::GetFullClassName(ClassID classId, tstring& fullClassName)
+{
+    HRESULT hr = S_OK;
+
+    ModuleID moduleId;
+    mdTypeDef typeDefToken;
+    IfFailLogRet(_corProfilerInfo->GetClassIDInfo2(
```

Comment:
This may fail in cases where the failure has to be gracefully handled, such as Incomplete data, array types, or composite types.

Reply:
See https://github.com/dotnet/samples/blob/2cf486af936261b04a438ea44779cdc26c613f98/core/profiling/stacksampling/src/sampler.cpp#L217

Reply:
Talked offline... @wiktork has near-future changes that will handle these particular cases, so I will decline to address them now.